### DISCUSSION TITLE
1.26.0-alpine Docker image fails to start

### DISCUSSION THREAD

**Discussion by jheiselman** [dani-garcia/vaultwarden#2854] [Q&A]
### Subject of the issue
Image version for vaultwarden/alpine:latest fails to start with error logged "Command not found: /start.sh".

### Deployment environment

* vaultwarden version: 1.26.0

* Install method: Docker image vaultwarden/alpine:latest as of 20 October, 2022

* Clients used: NA

* Reverse proxy and version: NA

* MySQL/MariaDB or PostgreSQL version: NA

* Other relevant details:

### Steps to reproduce
docker run vaultwarden/alpine:1.26.0

### Expected behaviour
Service should startup

### Actual behaviour
Service fails to start with error "Command not found: /start.sh"

### Troubleshooting data
Reverting to 1.25.2-alpine works properly

**Comment by jheiselman** [comment]
For what it's worth, the /start.sh script does appear to be in the image and indeed runs when called directly.

```shell
pi@raspberrypi:~/bitwarden $ docker run -it --rm --entrypoint bash vaultwarden/server:alpine
bash-5.1# ls
bin             dev             healthcheck.sh  lib             mnt             proc            run             srv             sys             usr             vaultwarden
data            etc             home            media           opt             root            sbin            start.sh        tmp             var             web-vault
bash-5.1# ls -l
total 39036
drwxr-xr-x    1 root     root          4096 Oct 14  2022 bin
drwxr-xr-x    2 root     root          4096 Oct 14  2022 data
drwxr-xr-x    5 root     root           340 Oct 21  2022 dev
drwxr-xr-x    1 root     root          4096 Oct 21  2022 etc
-rwxr-xr-x    1 root     root          1684 Oct 14  2022 healthcheck.sh
drwxr-xr-x    2 root     root          4096 Jul 18  2022 home
drwxr-xr-x    1 root     root          4096 Aug  5  2022 lib
drwxr-xr-x    5 root     root          4096 Jul 18  2022 media
drwxr-xr-x    2 root     root          4096 Jul 18  2022 mnt
drwxr-xr-x    2 root     root          4096 Jul 18  2022 opt
dr-xr-xr-x  232 root     root             0 Oct 21  2022 proc
drwx------    2 root     root          4096 Jul 18  2022 root
drwxr-xr-x    1 root     root          4096 Aug  5  2022 run
drwxr-xr-x    1 root     root          4096 Aug  5  2022 sbin
drwxr-xr-x    2 root     root          4096 Jul 18  2022 srv
-rwxr-xr-x    1 root     root           696 Oct 14  2022 start.sh
dr-xr-xr-x   12 root     root             0 Oct 21  2022 sys
drwxrwxrwt    2 root     root          4096 Jul 18  2022 tmp
drwxr-xr-x    1 root     root          4096 Aug  5  2022 usr
drwxr-xr-x    1 root     root          4096 Jul 18  2022 var
-rwxr-xr-x    1 root     root      39898984 Oct 14  2022 vaultwarden
drwxr-xr-x    9 root     root          4096 Oct 14  2022 web-vault
bash-5.1# cat start.sh
#!/bin/sh

if [ -r /etc/vaultwarden.sh ]; then
    . /etc/vaultwarden.sh
elif [ -r /etc/bitwarden_rs.sh ]; then
    echo "### You are using the old /etc/bitwarden_rs.sh script, please migrate to /etc/vaultwarden.sh ###"
    . /etc/bitwarden_rs.sh
fi

if [ -d /etc/vaultwarden.d ]; then
    for f in /etc/vaultwarden.d/*.sh; do
        if [ -r "${f}" ]; then
            . "${f}"
        fi
    done
elif [ -d /etc/bitwarden_rs.d ]; then
    echo "### You are using the old /etc/bitwarden_rs.d script directory, please migrate to /etc/vaultwarden.d ###"
    for f in /etc/bitwarden_rs.d/*.sh; do
        if [ -r "${f}" ]; then
            . "${f}"
        fi
    done
fi

exec /vaultwarden "${@}"
bash-5.1# ./start.sh
/--------------------------------------------------------------------\
|                        Starting Vaultwarden                        |
|                           Version 1.26.0                           |
|--------------------------------------------------------------------|
| This is an *unofficial* Bitwarden implementation, DO NOT use the   |
| official channels to report bugs/features, regardless of client.   |
| Send usage/configuration questions or feature requests to:         |
|   https://vaultwarden.discourse.group/                             |
| Report suspected bugs/issues in the software itself at:            |
|   https://github.com/dani-garcia/vaultwarden/issues/new            |
\--------------------------------------------------------------------/

[INFO] No .env file found.

[2022-10-21 18:36:05.839][vaultwarden][ERROR] No persistent volume!
########################################################################################
# It looks like you did not configure a persistent volume!                             #
# This will result in permanent data loss when the container is removed or updated!    #
# If you really want to use volatile storage set `I_REALLY_WANT_VOLATILE_STORAGE=true` #
########################################################################################

bash-5.1#
exit
```

**Comment by BlackDex** [comment]
See #2656

**Comment by jheiselman** [comment]
Okay, so a bind mount is no longer acceptable. That's fine, however, documentation needs to be updated.
https://github.com/dani-garcia/vaultwarden/wiki/Starting-a-Container still guides users to created a bind mount.

Further, the volume check that is mentioned as fixed in 1.25.2 apparently isn't as I am currently running 1.25.2-alpine with a bind mount without issue.

**Comment by BlackDex** [comment]
A bind mount is fine, that is what I'm using my self and no need for that env value.

If you can post your docker/docker-compose config, maybe we can help spotting something wrong.

**Reply by BlackDex** [reply]
Also, what does the following command return
```bash
docker exec -it vaultwarden cat /proc/1/mounts | grep "/data"
# Or within the container when exec to bash or sh
cat /proc/1/mounts | grep "/data"
```

**Reply by jheiselman** [reply]
I only just now saw this.

```
$ cat docker-compose.yml
version: "3.8"
services:
  bitwarden:
    image: vaultwarden/server:1.25.2-alpine
    ports:
      - 8080:80
      - 3012:3012
    restart: unless-stopped
    volumes:
      - ./data:/data/
      - ./.env:/.env
    extra_hosts:
      - "smtp.gmail.com:209.85.232.109"
```

Obviously, the version is pinned to 1.25.2 for me.

```
$ docker compose exec bitwarden cat /proc/1/mounts | grep "/data"
/dev/root /data ext4 rw,noatime 0 0
```

**Comment by nikoasr** [comment]
Hi, I fixed using  --entrypoint /start.sh, not really sure why only alpine:latest gives "Command not found: /start.sh" while all the others runs ok.

**Reply by BlackDex** [reply]
@nikoasr make sure you do a `docker pull vaultwarden:alpine`, else you will keep using an older version. On older versions it used dumb-init, mostly because of an issue with the base-images for some architectures. Those issues have been addressed by the base-image builders and thus no need for dumb-init anymore. You shouldn't have to change the entrypoint or cmd at all. It should just run with `docker run vaultwarden/server:alpine`

**Reply by nikoasr** [reply]
Hi BlackDex I already did `docker pull vaultwarden/server:alpine` and I pruned also all the images and redownload but the error is still there, it's like it cannot execute /start.sh but there is the file in /, I really don't understand. The version before works too.

**Reply by BlackDex** [reply]
btw, i do not understand why you set the entrypoint to that, since the cmd is already set to it.
That could cause an issue of course. Just do not configure entrypoint and cmd. All is fine from the image it self.

**Reply by nikoasr** [reply]
I tried on another ubuntu system and it works so the problem is only with my raspberry pi where the docker halt and if I see the logs it says: "Command not found: /start.sh", maybe some problem with the os/arch linux/arm/v7.

**Reply by BlackDex** [reply]
I tried this on my raspberry, and it started just fine.
This is a arm/v7 also.

**Reply by jheiselman** [reply]
My raspberry pi is where it fails as well. armv7

**Reply by BlackDex** [reply]
@jheiselman and @nikoasr could you answer the questions below?

What is the exact command you run to start the container?

What is the host os and version?

Which version of docker are you running?

Output of `docker images vaultwarden/*`

**Reply by nikoasr** [reply]
`docker run -d --name vaultwarden vaultwarden/server:alpine` (I removed all the options that I use from the startup command but still fails)
`Docker version 20.10.21`
`Raspbian GNU/Linux 10 (buster)`
````
REPOSITORY           TAG       IMAGE ID       CREATED       SIZE
vaultwarden/server   alpine    a094d192dff1   2 weeks ago   106MB
```

**Reply by BlackDex** [reply]
Strange, if i run this on my Pi like this

```bash
docker run --rm -it -e RUST_BACKTRACE=1 -e LOG_LEVEL=debug -e I_REALLY_WANT_VOLATILE_STORAGE=true -p8181:80 vaultwarden/server:alpine
```

It starts just fine.

It is the same exact image:
```
REPOSITORY           TAG       IMAGE ID       CREATED       SIZE
vaultwarden/server   alpine    a094d192dff1   2 weeks ago   106MB
```

Only difference i see right now is that I'm running a newer version of debian:
```
lsb_release -a
No LSB modules are available.
Distributor ID:	Raspbian
Description:	Raspbian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye
```

Also the exact same docker version it looks like:
```
docker version
Client: Docker Engine - Community
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 17:59:27 2022
 OS/Arch:           linux/arm
```

**Reply by jheiselman** [reply]
I use docker compose. Here is my docker-compose.yml. Obviously, I am now pinned to 1.25.2-alpine, but the issue occurred when I was pointed at `:alpine` which was, and currently is, 1.26.0.

```
version: "3.8"
services:
  bitwarden:
    image: vaultwarden/server:1.25.2-alpine
    ports:
      - 8080:80
      - 3012:3012
    restart: unless-stopped
    volumes:
      - ./data:/data/
    extra_hosts:
      - "smtp.gmail.com:209.85.232.109"
    environment:
      SIGNUPS_ALLOWED: ${SIGNUPS_ALLOWED}
      SMTP_HOST: ${SMTP_HOST}
      SMTP_FROM: ${SMTP_FROM}
      SMTP_PORT: ${SMTP_PORT}
      SMTP_SSL: ${SMTP_SSL}
      SMTP_USERNAME: ${SMTP_USERNAME}
      SMTP_PASSWORD: ${SMTP_PASSWORD}
      DOMAIN: ${DOMAIN}
```

```
$ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

```
$ docker version
Client: Docker Engine - Community
 Version:           20.10.20
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        9fdeb9c
 Built:             Tue Oct 18 18:19:59 2022
 OS/Arch:           linux/arm
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.20
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       03df974
  Built:            Tue Oct 18 18:17:51 2022
  OS/Arch:          linux/arm
  Experimental:     true
 containerd:
  Version:          1.6.8
  GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

```
$ docker images vaultwarden/*
REPOSITORY           TAG             IMAGE ID       CREATED        SIZE
vaultwarden/server   alpine          a094d192dff1   2 weeks ago    106MB
vaultwarden/server   1.25.2-alpine   4545dcf344ab   3 months ago   108MB
```

**Reply by BlackDex** [reply]
Only thing i see in common and different is the debian version.
Not sure how I'm going to be able to test armv7 debian 10 as a host currently.

**Reply by BlackDex** [reply]
@jheiselman and @nikoasr could you try the testing tagged image please?
So `vaultwarden/server:testing-alpine`, and see if that solves the issue?

**Reply by nikoasr** [reply]
Still not working

**Reply by jheiselman** [reply]
Same.

```
bitwarden-bitwarden-1  | Command not found: /start.sh
bitwarden-bitwarden-1 exited with code 1
bitwarden-bitwarden-1 exited with code 0
bitwarden-bitwarden-1 exited with code 0
bitwarden-bitwarden-1 exited with code 1
```

**Reply by jheiselman** [reply]
I just tested the "fix" of specifying the entrypoint as /start.sh and it does indeed work for me.

*WORKING*
```
version: "3.8"
services:
  bitwarden:
    image: vaultwarden/server:alpine
    entrypoint:
      - /start.sh
    ports:
      - 8080:80
      - 3012:3012
    restart: unless-stopped
    volumes:
      - ./data:/data/
      - ./.env:/.env
    extra_hosts:
      - "smtp.gmail.com:209.85.232.109"
```

However, changing it to use command as the Dockerfile does causes it to fail again.

*FAILED*
```
version: "3.8"
services:
  bitwarden:
    image: vaultwarden/server:alpine
    command:
      - /start.sh
    ports:
      - 8080:80
      - 3012:3012
    restart: unless-stopped
    volumes:
      - ./data:/data/
      - ./.env:/.env
    extra_hosts:
      - "smtp.gmail.com:209.85.232.109"
 ```

I'm starting to think that this is a problem with the Docker engine on this platform.

**Reply by BlackDex** [reply]
May i ask why you add `entrypoint` or `cmd` at all?
That should not be needed at all.

**Reply by jheiselman** [reply]
Just for troubleshooting purposes. If you look at my earlier docker-compose.yml, it doesn't contain either of those. But when I saw that someone said that adding entrypoint to their `docker run` fixed their issue, it caused me to think about testing it.

**Reply by BlackDex** [reply]
Strange, it should just work as-is.

**Reply by jheiselman** [reply]
Alright, I think I've worked out the issue.

1.25.2 used dumb-init. 1.26.0 uses /usr/bin/entry.sh as the entrypoint. In the init() function in entry.sh, it runs `command -v /start.sh`. According to the bash documentation, `Only shell builtin commands or commands found by searching the PATH are executed.` /start.sh is not on the PATH, so `command` just exits with a exit code of 1 triggering the error message we are seeing about `Command not found`.

There are a couple of options that I can see:

1. Move/link start.sh to one of `/usr/local/sbin, /usr/local/bin, /usr/sbin, /usr/bin, /sbin, or /bin`. `/usr/local/bin` would probably be the best advised. 
2. Cease using /usr/bin/entry.sh, though this probably means going back to dumb-init or something else similar

**Comment by BlackDex** [comment]
As a workaround for now you could use either
```bash
docker run .... vaultwarden/server:alpine sh /start.sh
```

Or with docker-compose:
```yaml
command: ["sh", "/start.sh"]
```
